### DISCUSSION TITLE
smart-doc golang语言API文档生成提案

### DISCUSSION THREAD

**Discussion by shalousun** [TongchengOpenSource/smart-doc-golang#1] [Ideas]
首先go的web框架接口主要是在controller层没法直观直接解析返回什么，入参和java spring框架httprequest一样不能直观的通过反射知道具体的参数类型。因此只能用go-swagger那个方式，[smart-doc-golang](https://github.com/TongchengOpenSource/smart-doc-golang)主要功能，支持生成OpenAPI、增加postman、jmeter的生成，打通torna推送。

后续用户的使用：
- 用户可以利用一下大模型插件，在开发工具上安装一些辅助开发的插件，生成go-swagger的注释tag。
- 用户安装smart-doc-golang。smart-doc-golang执行用户的命令例如生成OpenAPI、Postman等

首次PR实现目标：
1. 支持用户使用smart-doc-golang openapi指令生成OpenAPI 3.0文档, 只需要实现生成json或者yaml文件即可
2. 更改README.md文档添加smart-doc-golang的用户操作文档。
3. pr和注释采用英文提交，翻译可以使用千问或者其他大模型工具完成，避免后期再翻译修改

后期规划：
1. smart-doc-golang提取注释信息数据生成postman collection，用户可以导入Postman做测试
2. smart-doc-golang支持将提取的注释文档数据渲染成markdown文档
4. smart-doc-golang支持将提取的注释和API数据转换成Jmeter脚本。
5. smart-doc-golang 对接torna，实现将文档直接推送到统一的文档管理中心。

推广计划：
1. 完成对openapi 3.0的支持开启国内宣传，由社区PMC @shalousun 撰写文章发布宣传
2. openapi 3.0在国内测试成熟开启国外的宣传，由社区PMC @shalousun 撰写文章发布宣传

参考项目和文章：

参考：
https://github.com/go-swagger/go-swagger
https://github.com/swaggo/swag

文章：
https://medium.com/@zhengfke/go-go-swagger-generation-and-convert-to-openapi-3-0-3-5f5f8840f7d5
https://medium.com/@zhengfke/go-swagger-v2-convert-to-openapi-v3-cli-swag2op-5b3af29aec99

社区奖励：
1. 初期参与从0到1建设，提交大的模块功能性pr的同学，pr合并进入发版后，基本锁定社区Committer角色，
Committer角色获得者由@shalousun负责协调社区礼品和宣传。

**Comment by yann-y** [comment]
貌似是个不错的解决方案，还可以兼容go-swagger。

**Reply by shalousun** [reply]
对的，主要还是受限于go 框架写法问题，我们主要就是去兼容go-swagger的语法，然后做扩展，做生态，更适合企业落地的方案

**Comment by shalousun** [comment]
目前社区基于swaggo/swag实现是可以，但是由于swaggo/swag只支持swagger 2.0协议，对smart-doc社区来说需要做一些改变，主要的改造点如下：

1. 增加一个生成openapi 3.0协议的指令，smart-doc-golang去实现这个3.0协议的生成和指令
2. 增加一个生成postman的协议的指令，用于生成postman
3. 增加一个生成jemeter性能测试脚本的指令，用于生成jmeter脚本

pr提交时请先参考[社区贡献指](https://smart-doc-group.github.io/#/zh-cn/community/contributing)

**Reply by shalousun** [reply]
主流开源：
https://github.com/go-swagger/go-swagger
https://github.com/swaggo/swag
一些使用案例：
- https://medium.com/@zhengfke/go-go-swagger-generation-and-convert-to-openapi-3-0-3-5f5f8840f7d5
- https://medium.com/@zhengfke/go-swagger-v2-convert-to-openapi-v3-cli-swag2op-5b3af29aec99
